[docs] Add documentation for Observe (#44816)
expo/expo: docs/pages/eas/observe/get-started.mdx

---
title: Set up EAS Observe
sidebar_title: Get started
description: Learn how to install EAS Observe and start collecting performance metrics from your production app.
searchRank: 9
---

import { Collapsible } from '~/ui/components/Collapsible';
import { Terminal } from '~/ui/components/Snippet';
import { Step } from '~/ui/components/Step';

> EAS Observe tracks your app's startup performance in production. This guide walks you through installation, setup, and viewing your first metrics.

Reviewer: ```suggestion
EAS Observe tracks your app's startup performance in production. This guide walks you through installing the library, setting up your app, and viewing your first metrics.
```

Let's either use gerund phrases or nouns.

> ## Prerequisites

Reviewer: Can we please Prerequisites component here instead of collapsible component?

Component: https://github.com/expo/expo/blob/d8ed0817d4fd4e1ab2b4c3b76ca13c29bcb44820/docs/ui/components/Prerequisites/index.tsx#L1

Author: Oops, I forgot we had it! Maybe we should update the rest of the docs as well (most places with prerequisites don't use the component).

Reviewer: No problem and thank you for reminding me. I've created a task here: https://linear.app/expo/issue/ENG-20777/migrate-docs-pages-with-prerequisites-sections-to-the-prerequisites and will follow up today to update those pages

<Collapsible summary="An Expo user account">

EAS Observe is available to anyone with an Expo account. You can sign up at [expo.dev/signup](https://expo.dev/signup).

</Collapsible>

<Collapsible summary="Expo SDK 55 or later">

EAS Observe requires SDK 55 or later. Run `npx expo-doctor` to check your SDK version and `npx expo install --fix` to update dependencies.

</Collapsible>

<Collapsible summary="An EAS project">

Your app must be linked to an EAS project. Ensure `extra.eas.projectId` in your app config includes the project ID, or create one by running `eas init`.

</Collapsible>

<Step label="1">

## Install the library

Make sure you are using the latest version of `expo`, then install `expo-observe`:

> <Terminal cmd={['$ npx expo install --fix', '$ npx expo install expo-observe']} />

Reviewer: ```suggestion
<Terminal cmd={['$ npx expo install --fix', '', '$ npx expo install expo-observe']} />
```

I think there should be a line separate between these two commands.

</Step>

<Step label="2">

## Wrap your root layout

> Wrap your root layout with `AppMetricsRoot`. This higher-order component (HOC) automatically measures Time to First Render for you.

Reviewer: ```suggestion
Wrap your root layout with `AppMetricsRoot`. This higher-order component (HOC) automatically measures [Time to First Render](/eas/observe/reference/metrics/#time-to-first-render) for you.
```

> Wrap your root layout with `AppMetricsRoot`. This higher-order component (HOC) automatically measures Time to First Render for you.

Reviewer: ```suggestion
Wrap your root layout with `AppMetricsRoot`. This higher-order component (HOC) automatically measures Time to First Render (TTFR) for you.
```

We use TTI below, can we abbreviate here as well?

> ```jsx

Reviewer: ````suggestion
```tsx src/app/_layout.tsx
````

I think we should provide file title just for visibility's sake but up to you.

import { AppMetricsRoot } from 'expo-observe';

function RootLayout() {
  return (

>     // your layout

Reviewer: ```suggestion
    {/* your layout */}
```

JSX comment.

  );
}

export default AppMetricsRoot.wrap(RootLayout);
```

</Step>

<Step label="3">

## Mark interactive

Call `markInteractive()` when your app is fully ready for user interaction. This should be called **after** any initialization work behind the splash screen completes, such as:

- Checking for updates
- Authenticating the user
- Fetching initial data
- Animating the splash screen

> ```jsx

Reviewer: ````suggestion
```tsx
````

import { useEffect, useState } from 'react';
import * as SplashScreen from 'expo-splash-screen';
import { AppMetrics } from 'expo-observe';

> export default function App() {

Reviewer: I'd suggesting using the Expo Router example pattern like this one: https://docs.expo.dev/versions/latest/sdk/splash-screen/#delay-hiding-the-splash-screen

  const [isReady, setIsReady] = useState(false);

  useEffect(() => {
    async function prepare() {
      await authenticateUser();
      await fetchInitialData();
      setIsReady(true);
    }
    prepare();
  }, []);

  useEffect(() => {
    if (isReady) {
      SplashScreen.hide();
      AppMetrics.markInteractive();
    }
  }, [isReady]);

  // ...
}
```

> > **info** `markInteractive()` can safely be called multiple times per session, only the first call records the measurement. If your app has multiple entry screens (for example, an onboarding flow, login flow, or deep link targets), call `markInteractive` on **every one of these screens**. If you only place it on one screen, TTI will not be recorded when the app opens via a deep link to a different screen.

Reviewer: ```suggestion
> **info** `markInteractive()` can safely be called multiple times per session, but only the first call records the measurement. If your app has multiple entry screens (for example, an onboarding flow, login flow, or deep link targets), call `markInteractive` on **every one of these screens**. If you only place it on one screen, [Time to Interactive (TTI)](/eas/observe/reference/metrics/#time-to-interactive) will not be recorded when the app opens via a deep link to a different screen.
```

</Step>

<Step label="4">

## Create a new build

After installing `expo-observe` and adding the instrumentation, create a new build of your app:

<Terminal cmd={['$ eas build']} />

> **info** By default, metrics are not dispatched when `NODE_ENV=development`. To test your integration during development, see [Enable metrics in development](/eas/observe/configuration/#enable-metrics-in-development).

</Step>

<Step label="5">

## View your metrics

> In the [EAS dashboard](https://expo.dev), open your project and select **Observe** from the navigation menu to view metrics from your app.

Reviewer: ```suggestion
Open your project and open [**Observe** tab in EAS dashboard](https://expo.dev/accounts/[account]/projects/[project]/observe) to view metrics from your app:
```

Somtime back, Brent had feedback that we should provide direct links to dashboard (when we can) when guiding readers to EAS dashboard from our docs (I will add the writing style guide item for this and review our current docs). I think this place is a good opportunity.

```tsx app/_layout.tsx
import { Stack } from 'expo-router';
import * as SplashScreen from 'expo-splash-screen';
import { AppMetrics, AppMetricsRoot } from 'expo-observe';
import { useEffect, useState } from 'react';

// Keep the splash screen visible while we fetch resources
SplashScreen.preventAutoHideAsync();

function RootLayout() {
  const [isReady, setIsReady] = useState(false);

  useEffect(() => {
    async function prepare() {
      try {
        await authenticateUser();
        await fetchInitialData();
      } catch (e) {
        console.warn(e);
      } finally {
        setIsReady(true);
      }
    }

    prepare();
  }, []);

  useEffect(() => {
    if (isReady) {
      SplashScreen.hide();
      AppMetrics.markInteractive();
    }
  }, [isReady]);

  if (!isReady) {
    return null;
  }

  return <Stack />;
}

export default AppMetricsRoot.wrap(RootLayout);
```

</Tab>

<Tab label="Without Expo Router">

```tsx App.tsx
import { useEffect, useState } from 'react';
import * as SplashScreen from 'expo-splash-screen';
import { AppMetrics } from 'expo-observe';

// Keep the splash screen visible while we fetch resources
SplashScreen.preventAutoHideAsync();

> export default function App() {

Reviewer 2: I think we should change this to 
```tsx
function App() {....} 
export default AppMetricsRoot.wrap(App)
```

    Expo Observe is available to anyone with an Expo account. You can sign up at
    [expo.dev/signup](https://expo.dev/signup).
  </Requirement>
  <Requirement number={2} title="Expo SDK 55 or later">
    Expo Observe requires SDK 55 or later. Run `npx expo-doctor` to check your SDK version and `npx
    expo install --fix` to update dependencies.
  </Requirement>
  <Requirement number={3} title="An EAS project">
    Your app must be linked to an EAS project. Ensure `extra.eas.projectId` in your app config
    includes the project ID, or create one by running `eas init`.
  </Requirement>
</Prerequisites>

<Step label="1">

## Install the library

Make sure you are using the latest version of `expo`, then install `expo-observe`:

<Terminal cmd={['$ npx expo install --fix', '', '$ npx expo install expo-observe']} />

</Step>

<Step label="2">

## Wrap your root layout

Wrap your root layout with `AppMetricsRoot`. This higher-order component (HOC) automatically measures [Time to First Render (TTR)](/eas/observe/reference/metrics/#time-to-first-render) for you.

```tsx app/_layout.tsx
import { AppMetricsRoot } from 'expo-observe';

function RootLayout() {

>   return {

Reviewer 3: [🚨 **critical**] This snippet doesn't compile — `return { /* your layout */ }` returns an object literal, not JSX. Either fill in a real example or change to `return null;` / `return <></>;` so it's at least a valid placeholder. Same shape as the snippet a few lines down in the With/Without Expo Router tabs, which uses real JSX.

</Step>

<Step label="4">

## Create a new build

After installing `expo-observe` and adding the instrumentation, create a new build of your app:

<Terminal cmd={['$ eas build']} />

> **info** By default, metrics are not dispatched when `NODE_ENV=development`. To test your integration during development, see [Enable metrics in development](/eas/observe/configuration/#enable-metrics-in-development).

</Step>

<Step label="5">

## View your metrics

Open your project and open [**Observe** tab in EAS dashboard](https://expo.dev/accounts/[account]/projects/[project]/observe) to view metrics from your app.

For details on filtering, release comparison, and session investigation, see the [Dashboard guide](/eas/observe/dashboard/).

You can also query metrics from the terminal using the EAS CLI:

> - `eas observe:versions`: List app versions with build and update details.

Reviewer 3: [💡 suggestion] Suggestion: linking to or briefly describing what each command outputs would help — particularly `eas observe:metrics` vs `eas observe:events` (aggregated vs per-event). "Run with `--help`" is fine as a fallback, but a one-liner above each command would save the reader a roundtrip to the CLI.
